event_buffer.go

```diff
@@ -0,0 +1,68 @@
+package statful
+
+import (
+	"bytes"
+	"encoding/json"
+	"sync"
+)
+
+type eventBuffer struct {
+	buffer     []Event
+	dryRun     bool
+	eventCount int
+	flushSize  int
+	mu         sync.Mutex
+
+	Logger Logger
+	Sender Sender
+}
+
+func (e *eventBuffer) Event(event Event) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.buffer = append(e.buffer, event)
+	e.eventCount++
+
+}
+
+func (e *eventBuffer) Flush() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
```

Comment:
In this case, you don't want to defer the unlock because it'd mean the lock on `e` would be blocked until `e.flushBuffers(events)` finishes.

It should unlock straight after `e.drainBuffers()`.